renderdoc/driver/metal/metal_manager.h

```diff
@@ -0,0 +1,168 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#pragma once
+
+#include "core/resource_manager.h"
+#include "metal_resources.h"
+#include "metal_types.h"
+
+struct MetalInitialContents
+{
+  MetalInitialContents()
+  {
+    RDCCOMPILE_ASSERT(std::is_standard_layout<MetalInitialContents>::value,
+                      "MetalInitialContents must be POD");
+    memset(this, 0, sizeof(*this));
+  }
+
+  MetalInitialContents(MetalResourceType t)
+  {
+    memset(this, 0, sizeof(*this));
+    type = t;
+  }
+
+  template <typename Configuration>
+  void Free(ResourceManager<Configuration> *rm)
+  {
+    RDCASSERT(false);
+  }
+  bytebuf resourceContents;
+
+  // for plain resources, we store the resource type
+  MetalResourceType type;
+};
+
+struct MetalResourceManagerConfiguration
+{
+  typedef WrappedMTLObject *WrappedResourceType;
+  typedef void *RealResourceType;
+  typedef MetalResourceRecord RecordType;
+  typedef MetalInitialContents InitialContentData;
+};
+
+class MetalResourceManager : public ResourceManager<MetalResourceManagerConfiguration>
+{
+public:
+  MetalResourceManager(CaptureState &state, WrappedMTLDevice *device)
+      : ResourceManager(state), m_WrappedMTLDevice(device)
+  {
+  }
+  void SetState(CaptureState state) { m_State = state; }
+  CaptureState GetState() { return m_State; }
+  ~MetalResourceManager() {}
+  void ClearWithoutReleasing()
+  {
+    // if any objects leaked past, it's no longer safe to delete them as we would
+    // be calling Shutdown() after the device that owns them is destroyed. Instead
+    // we just have to leak ourselves.
+    RDCASSERT(m_LiveResourceMap.empty());
+    RDCASSERT(m_InitialContents.empty());
+    RDCASSERT(m_ResourceRecords.empty());
+    RDCASSERT(m_CurrentResourceMap.empty());
+    RDCASSERT(m_WrapperMap.empty());
+
+    m_LiveResourceMap.clear();
+    m_InitialContents.clear();
+    m_ResourceRecords.clear();
+    m_CurrentResourceMap.clear();
+    m_WrapperMap.clear();
+  }
+
+  // ResourceManager interface
+  ResourceId GetID(WrappedMTLObject *res)
+  {
+    if(res == NULL)
+      return ResourceId();
+
+    return res->id;
+  }
+  // ResourceManager interface
+
+  // template magic voodoo to unwrap types
```

Comment:
You'll want these helpers at global scope so that Unwrap() and so on are available in any code, not just inside the resource manager.

Reply:
89308d24b1b1cb2ef82dcfc3bf20077f5891f196